pkg/spoke/auth/cache_auth.go

```diff
@@ -0,0 +1,114 @@
+package auth
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/klog/v2"
+	workapiv1 "open-cluster-management.io/api/work/v1"
+)
+
+// ExecuteAction is the action of executing the manifest work
+type ExecuteAction string
+
+const (
+	// ApplyAndDeleteAction represents applying(create/update) resource to the managed cluster,
+	// and is responsiable for deleting the resource
+	ApplyAndDeleteAction ExecuteAction = "ApplyAndDelete"
+	// ApplyNoDeleteAction represents only applying(create/update) resource to the managed cluster,
+	// but is not responsiable for deleting the resource
+	ApplyNoDeleteAction ExecuteAction = "ApplyNoDelete"
+)
+
+// SubjectAccessReviewCheckFn is a function to checks if the executor has permission to operate
+// the gvr resource by subjectaccessreview
+type SubjectAccessReviewCheckFn func(ctx context.Context, executor *workapiv1.ManifestWorkSubjectServiceAccount,
+	gvr schema.GroupVersionResource, namespace, name string, ownedByTheWork bool) error
+
+// GetSARCheckerFn returns the SubjectAccessReviewCheckFn
+func (v *sarCacheValidator) GetSARCheckerFn() SubjectAccessReviewCheckFn {
+	return v.validator.checkSubjectAccessReviews
+}
+
+// NewExecutorCacheValidator creates a sarCacheValidator
+func NewExecutorCacheValidator(validator *sarValidator, executorCaches *ExecutorCaches) *sarCacheValidator {
+	return &sarCacheValidator{
+		validator:      validator,
+		executorCaches: executorCaches,
+	}
+}
+
+type sarCacheValidator struct {
+	executorCaches *ExecutorCaches
+	validator      *sarValidator
+}
+
+func GetExecuteAction(ownedByTheWork bool) ExecuteAction {
+	if ownedByTheWork {
+		return ApplyAndDeleteAction
+	}
+	return ApplyNoDeleteAction
+}
+
+func OwnedByWork(action ExecuteAction) bool {
+	return action == ApplyAndDeleteAction
+}
+
+func (v *sarCacheValidator) Validate(ctx context.Context, executor *workapiv1.ManifestWorkExecutor,
+	gvr schema.GroupVersionResource, namespace, name string,
+	ownedByTheWork bool, obj *unstructured.Unstructured) error {
+	if executor == nil {
+		return nil
+	}
+
+	if err := v.validator.executorBasicCheck(executor); err != nil {
+		return err
+	}
+
+	sa := executor.Subject.ServiceAccount
+	executorKey := ExecutorKey(sa.Namespace, sa.Name)
+	dimension := Dimension{
+		Namespace:     namespace,
+		Name:          name,
+		Resource:      gvr.Resource,
+		Group:         gvr.Group,
+		Version:       gvr.Version,
+		ExecuteAction: GetExecuteAction(ownedByTheWork),
+	}
+
+	if allow, ok := v.executorCaches.Get(executorKey, dimension); !ok {
+		err := v.validator.checkSubjectAccessReviews(ctx, sa, gvr, namespace, name, ownedByTheWork)
+		UpdateSARCheckResultToCache(v.executorCaches, executorKey, dimension, err)
+		if err != nil {
+			return err
+		}
```

Comment:
move return v.validator.checkEscalation(ctx, sa, gvr, namespace, name, obj) to here just to make it clearer that it should only be called when there is not cache.

Reply:
Whether there is a cache or not, we need to call the checkEscalation. because whether it will escalate not only depends on the role/rolebinding but also on the content of the resource to be applied, which may change